swagger_server/itm/itm_scenario.py

```diff
@@ -1,73 +1,240 @@
+        target_state :State = self.isd.current_scene.state
+
+        # If the scene has no action mappings, then the scenario can end.
+        if self.isd.current_scene.action_mappings == []:
+            self.session.end_scenario()
+
+        # Log the scene change
+        print(f"--> Changing to scene index {self.isd.current_scene_index}.")
+        self.session.history.add_history(
+            "Change scene",
+            {"session_id": self.session.session_id,
+            "scenario_id": self.id,
+            "scene_index": self.isd.current_scene_index},
+            target_state.to_dict() if target_state else None
+        )
+
+        '''
+        Merge state from new scene into session.state.  Approach:
+        0. Abort if no state to merge
+        1. Always replace entire `characters` structure.
+        2. For `supplies`, add or update any specified supplies.
+        3. For everything else, replace any specified (non-None) values
+           3a. Lists are copied whole (e.g., `character_importance`, `aid_delay`, `threats`).
+        4. Clear hidden data (e.g., character vitals)
+        '''
+        # Rule 0: Abort if no state to merge
+        if not target_state:
+            current_state.characters = []
+            return
+
+        # Rule 1: Always replace entire `characters` structure.
+        current_state.characters = target_state.characters
+
+        # Rule 2: For `supplies`, add or update any specified supplies.
+        if target_state.supplies:
+            target_types = []
+            current_types = []
+            for target_supply in target_state.supplies:
+                target_types.append(target_supply.type)
+                for current_supply in current_state.supplies:
+                    current_types.append(current_supply.type)
+                    if target_supply.type == current_supply.type:
+                        current_supply.quantity = target_supply.quantity
+                        current_supply.reusable = target_supply.reusable
+            new_types = [new_type for new_type in target_types if new_type not in current_types]
+            new_supplies = [new_supply for new_supply in target_state.supplies if new_supply.type in new_types]
+            current_state.supplies.extend(new_supplies)
+
+        # Rule 3: For everything else, replace any specified (non-None) values.
+        # Lists are copied whole (e.g., `character_importance`, `aid_delay`, `threats`).
+
+        # Top-level unstructured.
+        if target_state.unstructured:
+            current_state.unstructured = target_state.unstructured
+
+        # Threat state
+        if target_state.threat_state:
+            target = target_state.threat_state
+            current = current_state.threat_state
+            if not current:
+                current = target
+            else:
+                current.unstructured=target.unstructured if target.unstructured else current.unstructured
```

Comment:
I know this is probably something a linter would handle if we had one, but the spaces not being around the "=" caught my eye.  Totally nitpicky.

Reply:
I like nit-picky.   ;)  Will fix in next commit.